### ISSUE TITLE
test_handle_interrupted? may expose improper handle_interrupt logic

### ISSUE THREAD

**Issue by headius** [jruby/jruby#5314]
The test `test_handle_interrupted?` may fail at [test/mri/ruby/test_thread.rb:862](https://github.com/jruby/jruby/blob/master/test/mri/ruby/test_thread.rb#L862) due to flaws in `Thread#handle_interrupt` logic. If the handling is allowing exceptions to be thrown that should be masked, it could trigger push of `:ng1` earlier in the test, causing a sporadic failure as seen in this job: https://travis-ci.org/jruby/jruby/jobs/430135641

I am quarantining this test until we can determine if it is a flawed test or a bug in `handle_interrupt`.

FWIW I have stepped through `handle_interrupt` and been unable to reproduce the issue locally, so it may involve more complicated timing than can be simulated in a debug session.

**Comment by headius** [comment]
Other tests have been masked for this reason; it seems there's a bug in `handle_interrupt` that should be fixed before 9.2.1.

**Comment by larskanis** [comment]
The reason for `ng1` in the test is probably that the interrupt mask is currently not inherited to new threads in JRuby and Truffleruby. In MRI it's done [here](https://github.com/ruby/ruby/blob/5c16b21f44c3f035c87ad7fcb3df868700960a39/thread.c#L805).

Here is a minimal test case. In JRuby and Truffleruby this aborts the thread, in MRI this print `:ok` after one second:
```ruby
Thread.handle_interrupt(RuntimeError=>:never){ Thread.new{ sleep 1; p :ok } }.raise
```

Hope this can be fixed in 9.2.1.

**Comment by headius** [comment]
I  fixed the case that @larskanis explained, and `test_handle_interrupted?` does appear to pass consistently locally. The other three excluded tests still don't pass, though, and need further investigation.
